### ISSUE TITLE
Add delay and smarter verification between node restarts

### ISSUE THREAD

**Issue by jahkeup** [bottlerocket-os/bottlerocket-update-operator#12]
**What I'd like:**

Dogswatch should add some delay between the restart of Nodes in a cluster. During this time, the Controller should check in with the Node that has been updated to confirm that it has come up healthy and that Workloads have returned to it. After this, the Controller should have a configurable duration used to delay between each Node restart.

**Comment by samuelkarp** [comment]
This seems potentially related to [before and after reboot checks](https://github.com/coreos/container-linux-update-operator/blob/master/doc/before-after-reboot-checks.md).

**Comment by anguslees** [comment]
Indeed.  In case you want concrete use cases for before/after reboot checks, I use them (with coreos/flatcar update operator) to delay until the rook/ceph cluster is healthy[1], and to signal to rook/ceph that the rook storage cluster should set the "noout" flag[2].  After reboot clears the noout flag and again blocks until cluster is healthy again.

[1] eg: Data is replicated sufficiently.  This signal is "global" and much more complex than what a single pod readinessProbe can represent, which is why it can't be just a PodDisruptionBudget.  A better implementation might only consider the redundancy of the data on "this" node.  In particular, a naive time "delay", or generic check that pods were running again (as suggested in the issue description) would not be sufficient here.

[2] noout means the node outage that is about to happen is expected to be brief, and rook should not start frantically re-replicating "lost" data onto new nodes.

This wasn't my idea at all, the standard rook docs for this are: https://rook.io/docs/rook/v1.4/container-linux.html

Having used this for a long time now, it works great.  What might not be obvious at first is that the reboot script itself is deployed as a daemonset limited to nodes with the "before-reboot" label. That means it automatically "finds" and installs itself only on the relevant nodes, and only for the relevant time, which is pretty neat.  Debugging the system when updates are not proceeding does require an understanding of the various state machine interactions though, of course.

I would expect very similar challenges exist for something like an elasticsearch cluster, where data replication is important and also not represented in the "health" of any specific container.  I agree this probably points to a missing feature in PodDisruptionBudget, since it _is_ still fundamentally a question of "is it ok to make $pod unavailable now".

**Comment by chancez** [comment]
I'm not sure about the best approach, but one of my use-cases is jupyterhub notebook pods. These pods can't be interrupted, but we regularly cull inactive/idle ones. I'd like to be able to cordon the node that needs updating, and wait for the notebook pods to be stopped (which could be a while) before allowing with the node to be rebooted. This might be done using a tool like https://github.com/planetlabs/draino, but the update-operator would need coordination.

**Comment by jahkeup** [comment]
Thanks for sharing your use case and laying out what your ideal operation would look like.

> This might be done using a tool like [planetlabs/draino](https://github.com/planetlabs/draino), but the update-operator would need coordination.

Draino looks very closely related to this problem space. The project appears to build on the [Kubernetes autoscaler](https://github.com/kubernetes/autoscaler) in order to accomplish its task. I'm curious what other projects are integrating with the autoscaler and what they use to [enhance the features provided](https://github.com/kubernetes/autoscaler/blob/a5ed2cc3fe0aabd92c7758e39f1a9c9fe3bd6505/cluster-autoscaler/FAQ.md). 

We'll likely check out both of these projects as the design is sketched out.

**Comment by mikn** [comment]
@cbgbt @bcressey We are going to hit this soon and I would like to work on an implementation. Do you guys have any more thoughts on this? Sadly Draino has gone unmaintained for a fair few years by now.
We will have a combination of the problems outlined here;
1. we want to allow very long drain times to allow for long-running tasks to finish before rescheduling
2. We want to conduct cluster-level health checks on systems like, for instance, Ceph to ensure that the node is actually fully up before taking down the next node
3. We want to attach actions to specific daemonsets when we intend to reboot them (i.e communicate to the service that this is a scheduled reboot)

**Comment by yeazelm** [comment]
Sorry for the delay @mikn but I think those are valid use cases for better delay and verification knobs in brupop. We would review a PR if you offered one!
